[TensorExpr] Add dtype to class Buf.

ZolotukhinM opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #36617 [TensorExpr] Add a constructor accepting a name_hint to class Buf.
* **#36611 [TensorExpr] Add dtype to class Buf.**

Currently Buf represents underlying storage but it didn't have dtype.
That resulted in specifying dtypes in different places and there was no
mechanism to enforce its consistency: e.g. one could've created a kFloat
expression and use a kInt buffer to store its result. Now we're
centralizing where the logic regarding the storage is located and we can
start enforcing semantics rules.

Follow-ups: we can merge Buffer and BufHandle classes as the former is
now a mere wrapper over the latter.

Differential Revision: [D21027356](https://our.internmc.facebook.com/intern/diff/D21027356)

Commit:
[TensorExpr] Add dtype to class Buf.

Currently Buf represents underlying storage but it didn't have dtype.
That resulted in specifying dtypes in different places and there was no
mechanism to enforce its consistency: e.g. one could've created a kFloat
expression and use a kInt buffer to store its result. Now we're
centralizing where the logic regarding the storage is located and we can
start enforcing semantics rules.

Follow-ups: we can merge Buffer and BufHandle classes as the former is
now a mere wrapper over the latter.

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add dtype to class Buf."


Currently Buf represents underlying storage but it didn't have dtype.
That resulted in specifying dtypes in different places and there was no
mechanism to enforce its consistency: e.g. one could've created a kFloat
expression and use a kInt buffer to store its result. Now we're
centralizing where the logic regarding the storage is located and we can
start enforcing semantics rules.

Follow-ups: we can merge Buffer and BufHandle classes as the former is
now a mere wrapper over the latter.

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add dtype to class Buf."


Currently Buf represents underlying storage but it didn't have dtype.
That resulted in specifying dtypes in different places and there was no
mechanism to enforce its consistency: e.g. one could've created a kFloat
expression and use a kInt buffer to store its result. Now we're
centralizing where the logic regarding the storage is located and we can
start enforcing semantics rules.

Follow-ups: we can merge Buffer and BufHandle classes as the former is
now a mere wrapper over the latter.

Differential Revision: [D21027356](https://our.internmc.facebook.com/intern/diff/D21027356)

[ghstack-poisoned]

zheng-xq commented on test/cpp/tensorexpr/test_expr.cpp:
```diff
@@ -389,9 +389,9 @@ void testExprDynamicShapeAdd() {
-    Buffer a(BufHandle("a", {n}), kFloat);
-    Buffer b(BufHandle("b", {n}), kFloat);
-    Buffer c(BufHandle("c", {n}), kFloat);
+    Buffer a(BufHandle("a", {n}, kFloat));
```
While we are at this, can we use switch to 

Buffer a("a", {n}, kFloat};

Commit:
rebase on "[TensorExpr] Add dtype to class Buf."


Currently Buf represents underlying storage but it didn't have dtype.
That resulted in specifying dtypes in different places and there was no
mechanism to enforce its consistency: e.g. one could've created a kFloat
expression and use a kInt buffer to store its result. Now we're
centralizing where the logic regarding the storage is located and we can
start enforcing semantics rules.

Follow-ups: we can merge Buffer and BufHandle classes as the former is
now a mere wrapper over the latter.

Differential Revision: [D21027356](https://our.internmc.facebook.com/intern/diff/D21027356)

[ghstack-poisoned]
